### ISSUE TITLE
Unkown OpenSSL Error

### ISSUE THREAD

**Issue by chriscalitz** [pyca/cryptography#2471]
So I'm getting the following errors on Ubuntu server with Python 2.7 installed in Virtualenv.

OpenSSL 1.0.1 14 Mar 2012
Python 2.7.3

It seems to be very intermitted and I'm yet to find the comman thing which causes this issue.

Any ideas?
```
Traceback (most recent call last): 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/django/core/handlers/base.py", line 92, in get_response 
response = middleware_method(request) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/django/middleware/locale.py", line 21, in process_request 
check_path = self.is_language_prefix_patterns_used() 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/django/middleware/locale.py", line 56, in is_language_prefix_patterns_used 
for url_pattern in get_resolver(None).url_patterns: 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/django/core/urlresolvers.py", line 347, in url_patterns 
patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/django/core/urlresolvers.py", line 342, in urlconf_module 
self._urlconf_module = import_module(self.urlconf_name) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/django/utils/importlib.py", line 35, in import_module 
__import__(name) 

File "/var/www/example.com/htdocs/community/urls.py", line 14, in <module> 
admin.autodiscover() 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/django/contrib/admin/__init__.py", line 29, in autodiscover 
import_module('%s.admin' % app) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/django/utils/importlib.py", line 35, in import_module 
__import__(name) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/djcelery/admin.py", line 295, in <module> 
class PeriodicTaskAdmin(admin.ModelAdmin): 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/djcelery/admin.py", line 297, in PeriodicTaskAdmin 
form = periodic_task_form() 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/djcelery/admin.py", line 251, in periodic_task_form 
current_app.loader.import_default_modules() 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/djcelery/loaders.py", line 140, in import_default_modules 
super(DjangoLoader, self).import_default_modules() 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/celery/loaders/base.py", line 121, in import_default_modules 
tuple(maybe_list(self.app.conf.CELERY_INCLUDE)) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/celery/loaders/base.py", line 103, in import_task_module 
return self.import_from_cwd(module) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/celery/loaders/base.py", line 112, in import_from_cwd 
package=package, 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/celery/utils/imports.py", line 101, in import_from_cwd 
return imp(module, package=package) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/celery/loaders/base.py", line 106, in import_module 
return importlib.import_module(module, package=package) 

File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module 
__import__(name) 

File "/var/www/example.com/htdocs/community/overview/tasks.py", line 23, in <module> 
from community.overview.views import savePeopleChanges 

File "/var/www/example.com/htdocs/community/overview/views.py", line 26, in <module> 
from community.gdrive.files import insert_file 

File "/var/www/example.com/htdocs/community/gdrive/files.py", line 4, in <module> 
from apiclient import errors 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/apiclient/__init__.py", line 19, in <module> 
from googleapiclient import discovery 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/googleapiclient/discovery.py", line 68, in <module> 
from oauth2client.client import GoogleCredentials 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/oauth2client/client.py", line 53, in <module> 
from oauth2client import crypt 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/oauth2client/crypt.py", line 45, in <module> 
from oauth2client._openssl_crypt import OpenSSLVerifier 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/oauth2client/_openssl_crypt.py", line 18, in <module> 
from OpenSSL import crypto 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/OpenSSL/__init__.py", line 8, in <module> 
from OpenSSL import rand, crypto, SSL 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/OpenSSL/rand.py", line 11, in <module> 
from OpenSSL._util import ( 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/OpenSSL/_util.py", line 6, in <module> 
from cryptography.hazmat.bindings.openssl.binding import Binding 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 182, in <module> 
Binding.init_static_locks() 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 139, in init_static_locks 
cls._ensure_ffi_initialized() 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 134, in _ensure_ffi_initialized 
cls._register_osrandom_engine() 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 99, in _register_osrandom_engine 
_openssl_assert(cls.lib, cls.lib.ERR_peek_error() == 0) 

File "/home/username/.virtualenvs/env/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 43, in _openssl_assert 
errors 

InternalError: Unknown OpenSSL error. Please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. 
```

**Comment by leonGTHB** [comment]
I just have the same problem  and spent a lot of time on it. later on, i just reinstall cryptography, sudo pip uninstall cryptography && pip install cryptography , then it's okay!
python2.7.10
OS X EI CAPTAIN

**Comment by chriscalitz** [comment]
I thought the same but tried this when we had issue with older version which solve it for most of sites using this codebase. The strange thing is we have several instances of the same code running on the same server using the same virtualenv but only this one produces the error.

**Comment by reaperhulk** [comment]
@chriscalitz I assume you're running cryptography 1.1 here? Does this reliably occur on the one instance? If so, what happens if you modify `/home/username/.virtualenvs/env/lib/python2.7/site-packages/OpenSSL/_util.py` to remove the line `binding.init_static_locks()`?

**Comment by chriscalitz** [comment]
@reaperhulk I'm using cryptography 1.1 but can't seem to find the line you are referring to in the file you mentioned. It seems to happen only with a set of sites we don't have behind a SSL certificate. Could this have an influence?

**Comment by reaperhulk** [comment]
@chriscalitz that line is from PyOpenSSL, not cryptography, but I just noticed that there's no released version of pyOpenSSL that has that call.

Unfortunately without a reliable reproducer we aren't any closer to solving this issue :(

**Comment by leonGTHB** [comment]
sudo pip install pyOpenSSL==0.14

**Comment by chriscalitz** [comment]
I have managed to recreate this issue on another instance of the same codebase. I edited my apache virtualhost file and moved our development site from being behind a SSL certificate to running on port 80 without SSL and that caused thesite to start getting the same issues. As soon as I switch it back to SSL the errors stop. So it seems to not work for me when it is not behind SSL. Any ideas?

**Comment by reaperhulk** [comment]
@chriscalitz If you could modify the `cryptography/hazmat/bindings/openssl/binding.py` in your installed version of cryptography to return the error along with that exception and then put it here that'd be extremely useful. There's a function called `_openssl_assert` that should be changed to look like this: https://github.com/pyca/cryptography/blob/master/src/cryptography/hazmat/bindings/openssl/binding.py#L36

Once you change it if you trigger the bug the exception should have more information in it that you can paste here. It's possible this bug is similar to #2468 (although clearly it's caused by a mechanism other than SIP since this is occurring on Linux).

**Comment by chriscalitz** [comment]
No problem @reaperhulk here you go:

InternalError: Unknown OpenSSL error. Please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([_OpenSSLError(code=151441516L, lib=9, func=109, reason=108), _OpenSSLError(code=151441516L, lib=9, func=109, reason=108)])

**Comment by reaperhulk** [comment]
Interesting. This error is `PEM_read_bio:no start line`.

Are you using `mod_wsgi` here? If so, what happens if you set `WSGIApplicationGroup %{GLOBAL}`?

**Comment by chriscalitz** [comment]
So I added the following to virtualhost file as we are using embedded mode instead of demon process:

WSGIDaemonProcess example.com processes=2 threads=15 display-name=%{GROUP}
WSGIApplicationGroup %{GLOBAL}

But still get this:

InternalError: Unknown OpenSSL error. Please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([_OpenSSLError(code=151441516L, lib=9, func=109, reason=108), _OpenSSLError(code=151441516L, lib=9, func=109, reason=108)])

**Comment by reaperhulk** [comment]
Hmm, so the error here says that it's trying to load a key and failing. The unfortunate part is that it's apparently occurring before the binding is actually initialized. This implies some interaction with subinterpreters, a feature `mod_wsgi` uses heavily (and is problematic due to it having separate Python namespaces but a single shared C namespace).

Do you see this problem if you drop your processes/threads down to 1?

**Comment by chriscalitz** [comment]
Changed it but still seem to be getting the same error. It is just strange that it happens when I remove SSL from virtualhost as I can't seem to understand why this is influencing the key load failing.

**Comment by reaperhulk** [comment]
When you say remove SSL from the vhost what do you mean? Just stripping the `SSLCertificateKeyFile` and other lines or more than that?

**Comment by chriscalitz** [comment]
So removing:

SSLEngine On
SSLCertificateFile /etc/apache2/ssl/*.example.com.cert
SSLCertificateKeyFile /etc/ssl/localcerts/*.example.com.key
SSLCACertificateFile /etc/apache2/ssl/Example.crt

and changing the port the website is accessed on from 443 to 80.

This seems to reproduce it on another instance.

**Comment by reaperhulk** [comment]
Crazy idea: what happens if you only remove SSLEngine On and change 443 to 80 (leave the file/keyfile/cacertificatefile directives)?

The reason I'm investigating this is because ultimately the OpenSSL library is shared between cryptography and mod_ssl. This means if they're running in the same process that if an error occurs in apache and it doesn't clear the stack we might see it pop up in cryptography. It's probably not the issue, but it's worth a quick check...

**Comment by chriscalitz** [comment]
Unfortunately it made no difference. I also tried that ;-)

**Comment by reaperhulk** [comment]
Damn :smile:. @chriscalitz I realize this is asking a lot but if you're capable of building a docker environment that replicates this without putting any proprietary code or sensitive info in it that would be hugely helpful in getting this fixed. I am probably going to need to experiment for awhile to figure this one out and I need an env that replicates it before I can do that.

**Comment by bmbouter** [comment]
@chriscalitz @reaperhulk One of my co-workers identified this thread as related to an issue I also have experienced.

I experienced this issue in the Pulp Vagrant environment which is very self-contained already. It can be setup with [these instructions](https://github.com/pulp/pulp/blob/master/docs/dev-guide/contributing/dev_setup.rst). I think you only have to clone pulp and pulp_rpm and you can ignore the rest. The only meaningful thing to mention is that it occurs when the pulp_rpm tests are run, but it does *not* occur when you just run the failing test.

I experienced it with:
* pulp/pulp at commit b82ef6999ba97271d4f953f55aba4b81255fd68b
* pulp/pulp_rpm at commit 08ceb8f97aabfde2f1dc870577e84b3da15e6fa6

Just connect to the vagrant environment and run ptests which will run all the tests. You'll see what I see which is:

```
======================================================================
ERROR: Failure: InternalError (Unknown OpenSSL error. Please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce t
his.)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/.virtualenvs/pulp/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/vagrant/.virtualenvs/pulp/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/vagrant/.virtualenvs/pulp/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/vagrant/devel/pulp/streamer/test/unit/streamer/test_config.py", line 3, in <module>
    from pulp.streamer.config import load_configuration, DEFAULT_VALUES
  File "/home/vagrant/devel/pulp/streamer/pulp/streamer/__init__.py", line 2, in <module>
    from pulp.streamer.server import Streamer, Responder, StreamerListener  # noqa
  File "/home/vagrant/devel/pulp/streamer/pulp/streamer/server.py", line 9, in <module>
    from twisted.internet import reactor
  File "/usr/lib/python2.7/site-packages/Twisted-15.5.0-py2.7-linux-x86_64.egg/twisted/internet/reactor.py", line 38, in <module>
    from twisted.internet import default
  File "/usr/lib/python2.7/site-packages/Twisted-15.5.0-py2.7-linux-x86_64.egg/twisted/internet/default.py", line 56, in <module>
    install = _getInstallFunction(platform)
  File "/usr/lib/python2.7/site-packages/Twisted-15.5.0-py2.7-linux-x86_64.egg/twisted/internet/default.py", line 44, in _getInstallFunction
    from twisted.internet.epollreactor import install
  File "/usr/lib/python2.7/site-packages/Twisted-15.5.0-py2.7-linux-x86_64.egg/twisted/internet/epollreactor.py", line 24, in <module>
    from twisted.internet import posixbase
  File "/usr/lib/python2.7/site-packages/Twisted-15.5.0-py2.7-linux-x86_64.egg/twisted/internet/posixbase.py", line 18, in <module>
    from twisted.internet import error, udp, tcp
  File "/usr/lib/python2.7/site-packages/Twisted-15.5.0-py2.7-linux-x86_64.egg/twisted/internet/tcp.py", line 29, in <module>
    from twisted.internet._newtls import (
  File "/usr/lib/python2.7/site-packages/Twisted-15.5.0-py2.7-linux-x86_64.egg/twisted/internet/_newtls.py", line 21, in <module>
    from twisted.protocols.tls import TLSMemoryBIOFactory, TLSMemoryBIOProtocol
  File "/usr/lib/python2.7/site-packages/Twisted-15.5.0-py2.7-linux-x86_64.egg/twisted/protocols/tls.py", line 41, in <module>
    from OpenSSL.SSL import Error, ZeroReturnError, WantReadError
  File "/usr/lib/python2.7/site-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import rand, crypto, SSL
  File "/usr/lib/python2.7/site-packages/OpenSSL/rand.py", line 11, in <module>
    from OpenSSL._util import (
  File "/usr/lib/python2.7/site-packages/OpenSSL/_util.py", line 7, in <module>
    binding = Binding() 
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 95, in __init__
    self._ensure_ffi_initialized()
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 127, in _ensure_ffi_initialized
    cls._register_osrandom_engine()
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 99, in _register_osrandom_engine
    _openssl_assert(cls.lib, cls.lib.ERR_peek_error() == 0)
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 43, in _openssl_assert
    errors
InternalError: Unknown OpenSSL error. Please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this.
```

**Comment by reaperhulk** [comment]
@bmbouter this doesn't appear to be a stack trace from the latest cryptography release. Would you mind trying with the latest (it should hopefully output a bit more information as part of that InternalError)

**Comment by dkliban** [comment]
What is the latest version? Is it available for F23?

**Comment by reaperhulk** [comment]
1.1.2 is the latest. I don't know what F23 ships with, but significant changes have been made to our initialization code to attempt to alleviate this issue in 1.1. I'd recommend pip installing the latest inside a virtualenv to check it out.

**Comment by jeremycline** [comment]
@dkliban 1.0.2 is the current version in F23. It looks like that's being bumped to 1.1 in F24: https://apps.fedoraproject.org/packages/python-cryptography/builds/

**Comment by dkliban** [comment]
@reaperhulk I removed the rpm for python-cryptography and attempted to pip install cryptography and got the following error during install: http://fpaste.org/304441/45090021/

**Comment by reaperhulk** [comment]
You're going to need `libffi-dev` and `libssl-dev` (or whatever Fedora calls those packages) to successfully compile the version from PyPI. Sorry, should have mentioned that.

**Comment by dkliban** [comment]
@reaperhulk I just figured that out thanks to @jeremycline ... now on to the test.

**Comment by dkliban** [comment]
@reaperhulk Here is the new traceback:
```
Traceback (most recent call last):
  File "/home/vagrant/.virtualenvs/pulp/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/vagrant/.virtualenvs/pulp/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/vagrant/.virtualenvs/pulp/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/vagrant/devel/pulp/streamer/test/unit/streamer/test_config.py", line 3, in <module>
    from pulp.streamer.config import load_configuration, DEFAULT_VALUES
  File "/home/vagrant/devel/pulp/streamer/pulp/streamer/__init__.py", line 2, in <module>
    from pulp.streamer.server import Streamer, Responder, StreamerListener  # noqa
  File "/home/vagrant/devel/pulp/streamer/pulp/streamer/server.py", line 9, in <module>
    from twisted.internet import reactor
  File "/usr/lib64/python2.7/site-packages/twisted/internet/reactor.py", line 38, in <module>
    from twisted.internet import default
  File "/usr/lib64/python2.7/site-packages/twisted/internet/default.py", line 56, in <module>
    install = _getInstallFunction(platform)
  File "/usr/lib64/python2.7/site-packages/twisted/internet/default.py", line 44, in _getInstallFunction
    from twisted.internet.epollreactor import install
  File "/usr/lib64/python2.7/site-packages/twisted/internet/epollreactor.py", line 24, in <module>
    from twisted.internet import posixbase
  File "/usr/lib64/python2.7/site-packages/twisted/internet/posixbase.py", line 24, in <module>
    from twisted.internet import error, udp, tcp
  File "/usr/lib64/python2.7/site-packages/twisted/internet/tcp.py", line 29, in <module>
    from twisted.internet._newtls import (
  File "/usr/lib64/python2.7/site-packages/twisted/internet/_newtls.py", line 21, in <module>
    from twisted.protocols.tls import TLSMemoryBIOFactory, TLSMemoryBIOProtocol
  File "/usr/lib64/python2.7/site-packages/twisted/protocols/tls.py", line 41, in <module>
    from OpenSSL.SSL import Error, ZeroReturnError, WantReadError
  File "/usr/lib/python2.7/site-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import rand, crypto, SSL
  File "/usr/lib/python2.7/site-packages/OpenSSL/rand.py", line 11, in <module>
    from OpenSSL._util import (
  File "/usr/lib/python2.7/site-packages/OpenSSL/_util.py", line 6, in <module>
    from cryptography.hazmat.bindings.openssl.binding import Binding
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 182, in <module>
    Binding.init_static_locks()
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 139, in init_static_locks
    cls._ensure_ffi_initialized()
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 134, in _ensure_ffi_initialized
    cls._register_osrandom_engine()
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 99, in _register_osrandom_engine
    _openssl_assert(cls.lib, cls.lib.ERR_peek_error() == 0)
  File "/usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 43, in _openssl_assert
    errors
InternalError: Unknown OpenSSL error. Please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([_OpenSSLError(code=67530866L, lib=4, func=103, reason=114)])
```

**Comment by reaperhulk** [comment]
The error on the stack is `'error:04067072:rsa routines:RSA_EAY_PUBLIC_DECRYPT:padding check failed'`. This implies something else has been doing something with OpenSSL within the process and didn't clean up the error stack, so when we try to initialize it fails...

This is not a real fix, but do things work as expected if you replace the line:

`_openssl_assert(cls.lib, cls.lib.ERR_peek_error() == 0)`

with

`cls.lib.ERR_clear_error()`?

**Comment by jeremycline** [comment]
@reaperhulk I am quite certain that our (Pulp's) issue is that a second library (m2crypto) failing to clean up the error stack. Thanks for the tip!

**Comment by alex** [comment]
Do we have reproducers for any of these issues, without those I'm not sure what we can do. It looks like there's a high probability that these issues are caused by other code in the same proess failing to clean up OpenSSL properly.

**Comment by jeremycline** [comment]
In our case I did confirm it was other code in the same process failing to clean up OpenSSL's error stack. 

It would be cool if the error contained some more information, though. Perhaps a list of everything on the OpenSSL error stack and a note that it could be a different library failing to use OpenSSL correctly?

**Comment by alex** [comment]
The error message should have the complete OpenSSL error stack -- agreed that the initialization callsite should have a message which indicates another lib might be at fault.

**Comment by reaperhulk** [comment]
We return the integer representation of the errors, but presumably that's of limited utility to most people. If we called `ERR_error_string` on the errors then users would have a better idea of what's going on.

**Comment by alex** [comment]
Yeah, stringifying `OpenSSLError` to include the full string seems great.

On Mon, Mar 7, 2016 at 10:25 AM, Paul Kehrer <notifications@github.com>
wrote:

> We return the integer representation of the errors, but presumably that's
> of limited utility to most people. If we called ERR_error_string on the
> errors then users would have a better idea of what's going on.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/issues/2471#issuecomment-193297425>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by reaperhulk** [comment]
I'm going to go ahead and close this as in general this error msg is either caused by an error stack that is not being cleaned up properly (outside of cryptography's control) or a mod_wsgi issue with subinterpreters (the workaround for which will be documented in #2816).

**Comment by dikonoor** [comment]
reaperhulk,

I use python2-cryptography version 1.3.1 with Apache httpd (mod_wsgi) on RHEL 7.* and I get this error whenever I restart httpd . I followed your suggestion (https://github.com/pyca/cryptography/issues/2471) to replace /usr/lib64/python2.7/site-packages/cryptography/hazmat/bindings/openssl/bindings.py (below line) after which I cannot reproduce the error. If I revert the change, the problem can be reproduced. 

_openssl_assert(cls.lib, cls.lib.ERR_peek_error() == 0)

with

cls.lib.ERR_clear_error()?

I see multiple bugs opened along this line. but I am not able to figure out if a fix for this problem is available in any of the latest versions. Please let me know.

**Comment by alex** [comment]
Yup, you should upgrade: https://cryptography.io/en/latest/faq/#starting-cryptography-using-mod-wsgi-produces-an-internalerror-during-a-call-in-register-osrandom-engine

**Comment by dikonoor** [comment]
alex, Thanks for your response. Will it be possible to let me know into which version this fix has gone in. I am looking for version 1.7.1 el7 rpm for RHEL, which I can't find anywhere. I use OpenStack ocata releases which works only with upper limit version of 1.7.1.

**Comment by reaperhulk** [comment]
The fix is present in 1.7+
